File: CalibrationNCapture/CalibrationNCapture/calibration.py
"""
@Description : None
@Copied from : Henrychur
@Time : 2023/1/17 
"""
import cv2
import numpy as np
import glob

#recale images, otherwise it would take quite a long time 
def rescaleImg(img, scale=1):
    width = int(img.shape[1]*scale)
    height = int(img.shape[0]*scale)
    dimensions = (width, height)
    return cv2.resize(img, dimensions, interpolation=cv2.INTER_AREA)
# -------------------------------------------- #
# 设置寻找亚像素⻆点的参数
# 采⽤的停⽌准则是最⼤循环次数30和最⼤误差容限0.001
# -------------------------------------------- #
criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 30, 0.001) # 阈值
# -------------------------------------------- #
# 棋盘格模板规格
# 代表⻓宽⻆点的个数，注意棋盘格的顶点不被算⼊其中
# -------------------------------------------- #
w = 8
h = 11
# -------------------------------------------- #
# 世界坐标系中的棋盘格点
# 例如(0,0,0), (1,0,0), (2,0,0) ....,(8,5,0)
# 去掉Z坐标，记为⼆维矩阵
# -------------------------------------------- #
objp = np.zeros((w*h,3), np.float32)
objp[:,:2] = np.mgrid[0:w,0:h].T.reshape(-1,2)
objp = objp*30.0 # 棋盘格每个ceil的边⻓ 18.1 mm //30.0mm
# -------------------------------------------- #
# 储存棋盘格⻆点的世界坐标和图像坐标对
# -------------------------------------------- #
objpoints = [] # 在世界坐标系中的三维点
imgpoints = [] # 在图像平⾯的⼆维点

# img = cv2.imread('/home/kael/Documents/pyfiles/calibration/chessImg/6cee87da3b667bf1056761e96f38d2d.jpg')
# # img = rescaleImg(img)
# gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
# cv2.imshow('g', gray)
# cv2.waitKey(0)

images = glob.glob('./saveImg_k/*.png') # 图像路径
cnt = 0
i = 0
for fname in images:
    i += 1
    img = cv2.imread(fname)
    img = rescaleImg(img)
    # 获取画⾯中⼼点
    # 获取图像的⻓宽
    h1, w1 = img.shape[0], img.shape[1]
    # print(h1, w1) #t
    gray = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
    print(gray.shape[0], gray.shape[1]) #t
    u, v = img.shape[:2]
    # 找到棋盘格⻆点
    ret, corners = cv2.findChessboardCorners(gray, (w,h),None)
    if ret ==True:
        print("-1")
    else:
        print("-0")
    # 如果找到⾜够点对，将其存储起来
    if ret == True:
        print("cnt: \t", cnt)
        cnt = cnt + 1
        # 在原⻆点的基础上寻找亚像素⻆点
        cv2.cornerSubPix(gray,corners,(11,11),(-1,-1),criteria)
        #追加进⼊世界三维点和平⾯⼆维点中
        objpoints.append(objp)
        imgpoints.append(corners)
        # 将⻆点在图像上显⽰
        cv2.drawChessboardCorners(img, (w,h), corners, ret)
        cv2.namedWindow('findCorners', cv2.WINDOW_NORMAL)
        cv2.resizeWindow('findCorners', 640, 480)
        cv2.imshow('findCorners',img)
        cv2.imwrite('./res_k/findCorners'+str(i)+'.png', img)
        cv2.waitKey(200)
        if cnt > 100:
        # 太多的图⽚会导致计算缓慢，所以最⼤计算100张图⽚的
            break
cv2.destroyAllWindows()
print('正在计算')
#标定
ret, mtx, dist, rvecs, tvecs = \
    cv2.calibrateCamera(objpoints, imgpoints, gray.shape[::-1], None, None)
print("ret:",ret )
print("mtx:\n",mtx) # 内参数矩阵
print("dist畸变值:\n",dist) # 畸变系数 distortion cofficients = (k_1
print("rvecs旋转（向量）外参:\n",rvecs) # 旋转向量
print("tvecs平移（向量）外参:\n",tvecs) # 平移向量
newcameramtx, roi = cv2.getOptimalNewCameraMatrix(mtx, dist, (u, v), 0, (u, v))
print('newcameramtx外参',newcameramtx)

cv2.destroyAllWindows()

File: CaptureNCalibration/calib/Calibration.cpp
/*
 * @Copyright:
 * @author: Kael_Kang
 * @File name: Calibration
 * @Version: v1.0
 * @Date: 2023.5
 * @LastEditTime: 
 * @LastEditors: 
 * @description: 
 */
#include <iostream>
#include <dirent.h>

#include "../Camera/include/GxCamera.h"

/* ------- choose the chessboard configuration ------- */ 
const TermCriteria criteria = TermCriteria(TermCriteria::MAX_ITER + TermCriteria::EPS, 30, 0.001 );

const int chess_w = 8;
const int chess_h = 11;
const int cell_w = 30;
const int cell_h = 30;

const string read_path = "/home/rm/Desktop/CaptureNCalibration/saveImg_k";
const string show_res_path = read_path + "/../res_k";
/* ------- end setting the chessboard configuration ------- */ 

// ------- define containers ---------
vector<Point3f > board;

vector<vector<Point3f > > obj_points;
vector<vector<Point2f > > img_points;

vector<string > paths;

// ---------- set template of one chessboard cell ----------
void setBoard(vector<Point3f >& board, int chess_w, int chess_h ){
    for(int i = 0; i < chess_h; i++ ){
        for(int j = 0; j < chess_w; j++ ){
            Point3f pt;
            pt.x = i * cell_w;
            pt.y = j * cell_h;
            pt.z = 0;
            board.push_back(pt);
        }
    }

    return;
}

// ---------- extract files names from a directory ------------
vector<string > getFileNames(const string& dir_name, bool is_subdir = false ){
    DIR* dir_entry;
    struct dirent* dir_path;
    vector<string > file_names;

    if((dir_entry = opendir(dir_name.c_str() ) ) != NULL ){
        while((dir_path = readdir(dir_entry ) ) != NULL ){
            // cout <<  dir_path->d_name <<endl;
            if(strcmp(".", dir_path->d_name ) == 0 || strcmp("..", dir_path->d_name ) == 0 ){
                continue;
            }
            if(dir_path->d_type != DT_DIR ){
                file_names.push_back(dir_name + "/" +dir_path->d_name );
            }
            if(is_subdir && dir_path->d_type == DT_DIR ){
                vector<string > sub_file_names = getFileNames(dir_name + "/" + dir_path->d_name, true );
                file_names.insert(file_names.begin(), sub_file_names.begin(), sub_file_names.end() );
            }
        }
    }
    
    closedir(dir_entry);

    return file_names;
}

int main(){
// -- set necessities -- 
    paths = getFileNames(read_path );
    setBoard(board, chess_w, chess_h );
 
// -- define counters --
    uint cnt = 0;
    uint i = 0;

// -- the size of chessboard images is expected to be identical --
    int img_w;
    int img_h;

// -- calibrating circular -- 
    for(string& path : paths ){
        // cout << path <<endl;
        i++;
        Mat img = imread(path);

        img_w = img.size().width;
        img_h = img.size().height;
        Mat gray;

        cvtColor(img, gray, COLOR_BGR2GRAY );

// -- show each gray image --                    
        // imshow("gray iamge", gray);
        // waitKey(0);
        // if(gray.type()==CV_8UC1 ){
        //     cout << gray.channels() << endl;
        // }

// -- find corners and sub pixels -- 
        vector<Point2f > corners;
        bool ret = findChessboardCorners(
            gray, Size(chess_w, chess_h ), corners , CALIB_CB_ADAPTIVE_THRESH  | CALIB_CB_NORMALIZE_IMAGE 
        );
 
        cout << "is_find - " << ret << endl;

        if(ret ){
            cout << cnt << endl;
            cnt++;

            cornerSubPix(gray, corners, Size(11,  11 ), Size(-1, -1 ), criteria );

            obj_points.push_back(board );
            img_points.push_back(corners );

// -- draw corners -- 
            drawChessboardCorners(img, Size(chess_w, chess_h ), corners, ret );
            // namedWindow("findCorners ", WINDOW_NORMAL );
            // resizeWindow("findCorners ", 640, 480 );

            imshow("findCorners ", img );
            imwrite(show_res_path + "/findCorners_" + to_string(cnt ) + ".png", img );

            waitKey(40);

// -- don't be too many --
            if(cnt > 100 ){
                break;
            }
        }
    }

    destroyAllWindows();

// get final matrix and distcoeffs --
    cout << "Calculating ... ... " << endl;

    Mat cameraMatrix = Mat(3, 3, CV_32FC1, Scalar(0, 0, 0 ) );
    Mat distCoeffs = Mat(1, 5, CV_32FC1, Scalar(0, 0, 0 ) );
    vector<Mat> rvecsMat;
    vector<Mat> tvecsMat;
    double _err = calibrateCamera(
        obj_points, img_points, Size(img_w, img_h ), cameraMatrix, distCoeffs, rvecsMat, tvecsMat //, CV_CALIB_FIX_K3 
    );

// -- copy coeffs into pnp yaml -- 
    cout << cameraMatrix << endl;
    cout << distCoeffs << endl;
     
    return 0;
}







